tests/deepsparse/test_threads.py

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import concurrent.futures
+import time
+
+import pytest
+from deepsparse import ThreadPool
+
+
+def _heavy_compute(compute_time=0.25):
+    time.sleep(compute_time)
+
+
+@pytest.fixture()
+def threadpool():
+    """ "
+    An auto-delete fixture that yields a ThreadPool object to test
+    """
+    yield ThreadPool()
+
+
+@pytest.mark.parametrize(
+    "compute_times",
+    [
+        [0.25, 0.5, 0.75, 1],
+        [0.02, 0.05, 0.075, 0.1],
+        [0.1, 0.05, 0.3],
+    ],
+)
+def test_submit_is_asynchronous(threadpool, compute_times):
```

Comment:
I mentioned this to George, but I'll mention it again. I've had mixed results with timing-based tests like these. I've too often run into scenarios where some CI machine that is running the tests is overworked and it throws off all my timing expectations and leads to spurious test failures at inconvenient times.

Some of this also tests that `ThreadPoolExecutor` does what it claims to do. Perhaps there's an opportunity to introduce a test seam and create an instance of the Pipeline that has a mock version of a `ThreadPoolExecutor` that can be used to check that `ThreadPoolExecutor.submit` is being called with the expected arguments in the expected scenarios and leave testing `ThreadPoolExecutor` to python. Loosening the requirements of what `Pipeline._threadpool` should be would be helpful in this regard.

Reply:
Also, some of this is totally going to depend on what machine the tests are being run on. If there are 10 threads then you may do fine on a quad-core processor, but on a single core processor you could find that the multi-threaded stuff runs more slowly than the synchronous stuff because of thread contention.

Reply:
Thanks!, changed the test just to look at submit times in async mode vs end to end inference in sync mode